src/Compilers/Core/Portable/Compilation/NullableContext.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+using System;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents the state of the nullable analysis at a specific point in a file.
+    /// </summary>
+    [Flags]
+    public enum NullableContext
+    {
+        /// <summary>
+        /// Nullable warnings and annotations are off at this location.
+        /// </summary>
+        Disabled = 0b00_00,
+
+        /// <summary>
+        /// The nullable warning state is inherited from the project default.
+        /// </summary>
+        /// <remarks>
+        /// The project default can change depending on the file type. Generated
+        /// files have nullable off by default, regardless of of the project-level
+        /// default setting.
+        /// </remarks>
+        WarningsContextInherited = 0b00_01,
+
+        /// <summary>
+        /// The nullable annotation tate is inherited from the project default.
+        /// </summary>
+        /// <remarks>
+        /// The project default can change depending on the file type. Generated
+        /// files have nullable off by default, regardless of of the project-level
+        /// default setting.
+        /// </remarks>
+        AnnotationsContextInherited = 0b00_10,
+
+        /// <summary>
+        /// The current state of both warnings and annotations are inherited from
+        /// the project default.
+        /// </summary>
+        /// <remarks>
+        /// This flag is set by default at the start of all files.
+        ///
+        /// The project default can change depending on the file type. Generated
+        /// files have nullable off by default, regardless of of the project-level
+        /// default setting.
+        /// </remarks>
+        ContextInherited = WarningsContextInherited | AnnotationsContextInherited,
```

Comment:
can there be intuitive *user facing* explanation for the need for all these different enum values?  as these aren't being used anywhere, i don't really have anything to look at to say "ah, i see why these are needed".  Can the use cases for all of these be clarified?

Thanks!

Reply:
It's for code refactoring scenarios. Imagine you're doing generate method where the file you're *in* is `#nullable enable`, and the file you're generating into isn't. We have two options:

1. Drop any annotations in the newly generated method.
2. Surround the newly generated method with a `#nullable enable/#nullable disable` or `#nullable enable/#nullable restore` pair so your new method is in a happy land. But how do we pick between `#nullable disable` and #nullable restore`? It depends on what was *before* in the file, and whether we were previously restoring to the project default or whether the user had explicitly written `#nullable disable` at the top of the file.

As far as picking between 1 and 2, as soon as this API is in we're gong to immediately implement 1 since that's the easiest, but we may give the user an option to pick between them or make this configurable in some way. There's arguments for both cases depending on workflow and how a team wants to migrate to using nullability.

Reply:
@CyrusNajmabadi, let's say that your project has nullable disabled globally. Then, you go to generate a method in some file that doesn't have an explicit `#nullable` flag, and the thing we want to generate should be nullable aware. We'd put a `#nullable enable` before the generated code. But after the code, what do we put? If we put `#nullable disable`, that later screws you when you say "Alright, time to flip nullability on for my entire project!", because the code in that file after the explicit disable doesn't get nullability. Conversely, if you had nullability turned on everywhere but had put a `#nullable disable` in the file, us putting a `#nullable restore` in your code would enable nullable for a portion of code that you didn't intend to enable it for. So to surround with the correct pragmas, code generates need to know both the state of the feature, and whether the feature was explicitly set to that state or was inheriting that from the project defaults.